src/transformStateWithClones.js

```diff
@@ -6,8 +6,58 @@
  *
  * @return {Object[]}
  */
+// function transformStateWithClones(state, actions) {
+//   const stateHistory = [];
+//   const stateCopy = { ...state };
+
+//   for (const action of actions) {
+//     if (action.type === 'addProperties') {
+//       Object.assign(stateCopy, action.extraData);
+//       stateHistory.push(Object.assign({}, stateCopy));
+//     } else if (action.type === 'removeProperties') {
+//       for (const key of action.keysToRemove) {
+//         delete stateCopy[key];
+//       }
+
+//       stateHistory.push(Object.assign({}, stateCopy));
+//     } else if (action.type === 'clear') {
+//       Object.keys(stateCopy).forEach((key) => delete stateCopy[key]);
+//       stateHistory.push(Object.assign({}, stateCopy));
+//     }
+//   }
+
+//   return stateHistory;
+// }
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  const stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateCopy, action.extraData);
+        stateHistory.push(Object.assign({}, stateCopy));
```

Comment:
The `stateHistory.push` operation should be moved outside of the switch block and placed at the end of each loop cycle. This ensures that the state is consistently recorded after each action, which is crucial for maintaining an accurate history.